### PR TITLE
[BE][1/n] simplify train.py and rename things for consistency

### PR DESCRIPTION
**Pull request by tianyu-l**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #496
* #495
* __->__ #494

Including:
- move TrainState from `train.py` to `checkpoint.py`
- create `optimizer.py` to include things related to optimizers and lr scheduler.
- rename `logging_utils.py` to `logging.py`
- unify various build prefixes (`build_`, `create_`, `get_`) to `build_`
- simplify utils import by doing `import torchtitan.utils as utils`
- move `get_metrics_rank` from `utils.py` to `metrics.py` to make `build_metric_logger` simpler
- create `GarbageCollection` in `utils.py` to hide gc details from `train.py`
- reorder definition and initialization of some objects in `train.py` to be closer to where they are first used
- expose `build_pipeline_schedule` to `torchtitan.parallelisms`
- other minor improvements to reduce the amount of import in `train.py`

Together with #496, import LoC in `train.py` drops from 51 to 23. Total LoC of `train.py` drops from 537 to 424

### CODE REVIEW

**Review by awgu — APPROVED** [review]
Thanks!
